### PR TITLE
Extend a component package

### PR DESCRIPTION
**Pull request by alexkilgour**
Extracts the functionality for extending a component from the [Frontend Package Manager](https://github.com/springernature/frontend-package-manager/) into its own package, so that it can be used elsewhere.

This replicates the existing functionality, whilst adding some additional options, and refactoring

### Files mapped to `frontend-package-manager`

* `_utils/_get-remote-file-list.js` => [`_utils/_get-extended-file-list.js`](https://github.com/springernature/frontend-package-manager/blob/master/lib/js/_utils/_get-extended-file-list.js)
* `_utils/_get-remote-file.js` => [`_utils/_get-remote-file.js`](https://github.com/springernature/frontend-package-manager/blob/master/lib/js/_utils/_get-remote-file.js)
* `_utils/_merge-packages.js` => [`_utils/_merge-extended-package.js`](https://github.com/springernature/frontend-package-manager/blob/master/lib/js/_utils/_merge-extended-package.js)

### Additions and changes

* Convert from `Promises` to `Async/Await`
* Add functionality to get details of the package extension
* Add functionality to extend into a different folder

### Test coverage

```
-------------------------------------|----------|----------|----------|----------|
File                                 |  % Stmts | % Branch |  % Funcs |  % Lines |
-------------------------------------|----------|----------|----------|----------|
 util-package-extender/lib/js        |      100 |      100 |      100 |      100 |
  index.js                           |      100 |      100 |      100 |      100 |
 util-package-extender/lib/js/_utils |      100 |      100 |      100 |      100 |
  _get-local-file-list.js            |      100 |      100 |      100 |      100 |
  _get-remote-file-list.js           |      100 |      100 |      100 |      100 |
  _get-remote-file.js                |      100 |      100 |      100 |      100 |
  _merge-packages.js                 |      100 |      100 |      100 |      100 |
-------------------------------------|----------|----------|----------|----------|
```

### CODE REVIEW

File: packages/util-package-extender/lib/js/_utils/_get-remote-file-list.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+function getFileList(json, filePaths = [], buildPath = '') {
+	json.forEach(file => {
+		if (file.type === 'file') {
+			// Add file to array
+			filePaths.push(`${buildPath}${file.name}`);
+		}
+		if (file.type === 'directory') {
+			// recursvely call function, adding directory to build path
+			getFileList(file.files, filePaths, `${buildPath}${file.name}/`);
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/lib/js/_utils/_get-remote-file-list.js]: +1 for a lovely recursive function!

File: packages/util-package-extender/lib/js/_utils/_merge-packages.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * Copy local file to outputDirectory
+ * @private
+ * @param {String} file name of the local file
+ * @param {String} sourcePath path to the local file
+ * @param {String} destinationPath output path of the file
+ * @return {Promise<Object>}
+ */
+async function mergeLocalFile(file, sourcePath, destinationPath) {
+	try {
+		await fs.copy(sourcePath, destinationPath);
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/lib/js/_utils/_merge-packages.js]: Not sure you want to but you can use `util.promisfy` to get promise versions of any callback based api. Save an extra install 

```
const fs = require('fs');
const {promisfy} = require('util');
const copy = promisify(fs.copyFile);
```

Comment (alexkilgour) [reply packages/util-package-extender/lib/js/_utils/_merge-packages.js]: You can! I always forget this 😄

Comment (alexkilgour) [reply packages/util-package-extender/lib/js/_utils/_merge-packages.js]: Just realised that I use `ensureDir` which isn't a standard node.fs command. Could rewrite it 🤔

Comment (alexkilgour) [reply packages/util-package-extender/lib/js/_utils/_merge-packages.js]: It's a fair bit more code to remove this dependency, so I'm leaning towards leaving it as is (remembering that this is why I didn't do this in the first place!)

File: packages/util-package-extender/lib/js/_utils/__mocks__/_get-local-file-list.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/**
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/lib/js/_utils/__mocks__/_get-local-file-list.js]: Why is this mock file in the main bit of code, and not under `__mocks__` somewhere in the `__tests__` directory?

File: packages/util-package-extender/lib/js/_utils/_get-remote-file.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ */
+function generateErrorMessage(error) {
+	switch (error.name) {
+		case 'RequestError':
+			return `request failed with ${error.code}`;
+		case 'HTTPError':
+		case 'MaxRedirectsError':
+			return `${error.statusCode} (${error.statusMessage})`;
+		default:
+			return `Uh-oh Something went wrong`;
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/lib/js/_utils/_get-remote-file.js]: Add the error.name
```suggestion
			return `Uh-oh Something went wrong. The unhandled error was ${error.name}`;
```

Also, inconsistent casing! :twitch:

File: packages/util-package-extender/lib/js/_utils/_get-remote-file.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * @param {String} url the request
+ * @param {Object} options got configuration, default: config
+ * @return {Promise<Object>}
+ */
+async function getRemoteFile(url, options = config) {
+	try {
+		const response = await got(url, options);
+		return response.body;
+	} catch (error) {
+		throw new Error(`${generateErrorMessage(error)}`);
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/lib/js/_utils/_get-remote-file.js]: You shouldn't need to interpolate a straight function call
```suggestion
		throw new Error(generateErrorMessage(error));
```

File: packages/util-package-extender/package.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "dependencies": {
+    "find-up": "^4.1.0",
+    "fs-extra": "^8.1.0",
+    "gitignore-globs": "^0.1.1",
+    "glob": "^7.1.4",
+    "got": "^9.6.0",
+    "nock": "^10.0.6"
```
Comment (joeyciechanowicz) [inline packages/util-package-extender/package.json]: This should be in the `devDependencies`

**Comment by alexkilgour** [comment]
> Will this util always be used in code, or does it need to expose a CLI interface?

This will always be used in code as part of two (at the moment) CLI interfaces, but never standalone

File: packages/util-package-extender/lib/js/_utils/__mocks__/_get-local-file-list.js (hunk 3 above)
Comment (alexkilgour) [reply packages/util-package-extender/lib/js/_utils/__mocks__/_get-local-file-list.js]: This needs to live in a `__mocks__` folder next to where the file is defined, so

```
lib/js/_utils/_get-local-file-list.js
lib/js/_utils/__mocks__/_get-local-file-list.js
```

Or at least that is how Jest used to function?

File: packages/util-package-extender/lib/js/_utils/_get-remote-file.js (hunk 5 above)
Comment (alexkilgour) [reply packages/util-package-extender/lib/js/_utils/_get-remote-file.js]: Good spot, this used to have custom text in the error as well

**Pull request merged**